packages/selected-network-controller/tests/SelectedNetworkMiddleware.test.ts

```diff
@@ -0,0 +1,103 @@
+import { ControllerMessenger } from '@metamask/base-controller';
+import type {
+  NetworkControllerEvents,
+  NetworkControllerActions,
+} from '@metamask/network-controller';
+
+import { buildSelectedNetworkControllerMessenger } from './utils';
+import { SelectedNetworkControllerActionTypes } from '../src/SelectedNetworkController';
+import createSelectedNetworkMiddleware from '../src/SelectedNetworkMiddleware';
+
+const buildMessenger = () => {
+  return new ControllerMessenger<
+    NetworkControllerActions,
+    NetworkControllerEvents
+  >();
+};
+
+const buildNetworkControllerMessenger = (messenger = buildMessenger()) => {
+  return messenger.getRestricted({
+    name: 'NetworkController',
+    allowedActions: ['NetworkController:getState'],
+    allowedEvents: ['NetworkController:stateChange'],
+  });
+};
+
+const noop = jest.fn();
+
+describe('createSelectedNetworkMiddleware', () => {
+  it('puts networkClientId on request', async () => {
+    const selectedNetworkControllerMessenger =
+      buildSelectedNetworkControllerMessenger();
+    const networkControllerMessenger = buildNetworkControllerMessenger();
+    const middleware = createSelectedNetworkMiddleware(
+      selectedNetworkControllerMessenger,
+      networkControllerMessenger,
+    );
+
+    const req = {
+      origin: 'example.com',
+    } as any;
+
+    const mockGetNetworkClientIdForDomain = jest
+      .fn()
+      .mockReturnValue('mockNetworkClientId');
+
+    selectedNetworkControllerMessenger.registerActionHandler(
+      SelectedNetworkControllerActionTypes.getNetworkClientIdForDomain,
+      mockGetNetworkClientIdForDomain,
+    );
+
+    await new Promise((resolve) => middleware(req, {} as any, resolve, noop));
```

Comment:
Instead of calling the middleware directly, what do you think about plugging it into a `json-rpc-engine` and having the engine call the middleware? Then we know that it doesn't cause issues when used in that context (since it will only ever be used in that context). See tests, for instance, for `eth-json-rpc-middleware`: https://github.com/MetaMask/eth-json-rpc-middleware/blob/main/src/block-cache.test.ts

Reply:
I think it makes sense to add an additional test which checks that this middleware does work when its used in json-rpc-engine, but I don't think every test should be using a real json-rpc-engine. doing so just makes testing more annoying. 1 single test that ensures it doesn't throw when its used as json-rpc-engine middleware should do the job.

Reply:
all done!